### DISCUSSION TITLE
Is there any way to export resulting field contours?

### DISCUSSION THREAD

**Discussion by Astrei** [bfieldtools/bfieldtools#52] [Q&A] [answered]
Hi again. Did you consider an opportunity to export the resulting field contours for further processing? I tried to do it via trimesh tools by applying trimesh.creation.sweep_polygon with controllable thickness and via vtk .stl export from the Mayavi scene but not succeed in both cases. Maybe you already have some way for this or some ideas? Thank you in advance.

**Comment by rzetter** [comment]
Hi! For now, the only contour export functionality integrated into bfieldtools is in the kicad_io module, which has a small function that write the files in the format of the KiCAD PCB design software. What specifically do you want to do with the contours, what kind of processing are you referring to? Are you looking to 3D print the contours with some conducting material, or "boolean-remove" the contours from a coil former/frame?

**Reply by Astrei** [reply]
My idea was to use the bfieldTools output in the shape of 3d wire structures further in COMSOL or other specific software to validate results and also estimate, for example, what would happen if displacement currents were not excluded, or to calculate the SAR in biological samples, etc. 
I also want to make a 3d printed coil holder with wire paths included for coil assembling. That is more simple than boolean extraction but also pretty useful. Now I am still trying to make it with the trimesh, if Ill succeed I'll share my results here.

**Comment by Astrei** [comment]
Ok, here's my solution, which seems to work. 'Loops' is the Lineconductor type output from the optimization. The code converts paths into tubes and saves them in a .obj file for further processing. Of course, the direction of the current will be lost, but that's another story.

```python
import trimesh
from shapely.geometry import Polygon

# Generate circular polygon
tube_radius = 0.001 #radius for the resulting tubes
n_components = 10 #number of the circle segments
vec = np.array([0.0, 1.0]) * tube_radius
angle = 2 * np.pi / n_components
rotmat = np.array([[np.cos(angle), -np.sin(angle)], [np.sin(angle), np.cos(angle)]])
perim = []
for _ in range(n_components):
    perim.append(vec)
    vec = np.dot(rotmat, vec)
poly = Polygon(perim)

scene = trimesh.Scene()
for contour_num in range(len(loops.paths)):
    scene.add_geometry(trimesh.creation.sweep_polygon(polygon=poly, path=loops.discrete[contour_num]))
trimesh.exchange.export.export_scene(scene, 'Mesh.obj', file_type='obj', resolver=None)
```

**Reply by Astrei** [reply, marked as answer]
And here is its improved version which generates .stl and SolidWorks 3d curve format:

```python
import trimesh
from shapely.geometry import Polygon

# Line smoothing algorithm, thanks to https://stackoverflow.com/questions/47068504/where-to-find-python-implementation-of-chaikins-corner-cutting-algorithm
def chaikins_corner_cutting(coords, refinements=5):
    coords = np.array(coords)

    for _ in range(refinements):
        L = coords.repeat(2, axis=0)
        R = np.empty_like(L)
        R[0] = L[0]
        R[2::2] = L[1:-1:2]
        R[1:-1:2] = L[2::2]
        R[-1] = L[-1]
        coords = L * 0.75 + R * 0.25

    return coords

#Fuse close-placed points
def fuse(line_data, distance_coeff):
    deletion_mask = np.zeros_like(line_data)
    for row_num in range(len(line_data) - 1):
        distance_coeff_check=(line_data[row_num][0]-line_data[row_num+1][0])**2 + (line_data[row_num][1]-line_data[row_num+1][1])**2 + (line_data[row_num][2]-line_data[row_num+1][2])**2
        if distance_coeff_check <= distance_coeff:
            deletion_mask[row_num] = [1,1,1]
    fused_data = np.delete(line_data, np.where(np.all(deletion_mask==1,axis=1)), 0)

    return fused_data


# Generate circular polygon and 3d scene
tube_radius = 0.001 #radius for the resulting tubes
poly = Polygon(np.array([(0, 0), (0, 1), (1, 1)])* tube_radius)
scene = trimesh.Scene()

for contour_num in range(len(loops.discrete)):
    curve_fused = fuse(loops.discrete[contour_num],1e-6)
    if not np.array_equal(curve_fused[[0]],curve_fused[[-1]], equal_nan=False):
        print('Closing the curve ' + str(contour_num))
        curve_fused = np.append(curve_fused,curve_fused[[0]], axis=0) # closing the contour loop by copying starting point to the end if fusion broke it
    curve = chaikins_corner_cutting(curve_fused, refinements=3)
    np.savetxt("C:\\Users\\Path to your folder\\" + str(contour_num) + ".sldcrv", curve, delimiter=",")
    try:
        mesh_out = trimesh.creation.sweep_polygon(polygon=poly, path=curve)
        scene.add_geometry(mesh_out)
    except:
        continue

trimesh.exchange.export.export_scene(scene, 'C:\\Users\\Path to your folder\\Mesh.stl', file_type='stl', resolver=None)
```
I added a smoothing algorithm and a basic distance check to remove redundant points in the contours. These are two big sources of problems for translating curves into solids. But even that does not guarantee that swept operation would be performed well, so the .stl output file is used only for a basic visual check of geometry. It turned out that it is much better to get only curves and use CAD for further 3d modeling (what a surprise).

Solidworks macro for multiple curves import (20 curves example):

```VBA

Sub main()

Set swApp = Application.SldWorks

Set Part = swApp.ActiveDoc
Dim myModelView As Object
Set myModelView = Part.ActiveView
For i = 0 To 19
    Part.InsertCurveFile ("C:\Users\Path to your folder\" & i & ".sldcrv")
Next i
End Sub
```
Once imported, I highly recommend redrawing your coils with 3D sketch curves. This will fix most problems with swept operations. During that, it is also easy to connect contours to form a single coil, change the number of points, and correct the geometry if needed. Most of the structures could be easily transferred to COMSOL afterward. Some fine wires could have problems with the mesh there, but that's case specific.

**Comment by Markuus55** [comment]
A bit late to the discussion, but the current direction can be saved when exporting the coil in a direction-sensitive file format. I have used dxf-files created using ezdxf to import coils as line currents in COMSOL Example code below, in which loops are a discretized coil:


```python
import ezdxf

def save_coil_loops(loops, save_path, file_name):

    document = ezdxf.new('R2010')
    modelspace = document.modelspace()

    for loop in loops.entities:
        loop_vert = loops.vertices[loop.points]
        modelspace.add_polyline3d(loop_vert, dxfattribs={'lineweight': 200})

    document.saveas(save_path + file_name)
```
